### PR TITLE
feat(idp): [Issue #45] idp server implementations

### LINKED ISSUE #45: [FEAT]: Implement OpenIddict IdP Server with MFA, Trusted Devices, and QR Login (.NET 10)
### 🔍 Problem Statement

Our current authentication infrastructure needs to be modernized to support enterprise-grade security flows. We require a dedicated Identity Provider (IdP) capable of handling complex state machines for authentication, including Multi-Factor Authentication (MFA), Trusted Device fingerprinting, and cross-device QR code logins. Relying on basic cookie authentication or outdated Identity frameworks will not scale with our security requirements. We need a robust, OIDC-compliant server that gives us complete control over token issuance, validation, and custom authentication grants without relying on external third-party infrastructure.

### ✨ Proposed Solution

Implement a dedicated Identity Provider (IdP) using **OpenIddict** running on **.NET 10**. This server will act as the single source of truth for user authentication and authorization. 

The IdP will support the following core OIDC and custom flows:
1.  **Standard Login/Logout:** Support for the Authorization Code Flow with PKCE. Complete session termination via a standard End Session endpoint.
2.  **Token Refresh:** Implementation of sliding-expiration Refresh Tokens with automatic rotation and revocation to prevent replay attacks.
3.  **MFA & Trusted Devices (Stateful Login):** A multi-step login pipeline. The IdP will evaluate incoming device fingerprints. If the device is recognized and trusted, standard tokens are issued. If untrusted, the IdP will issue a "partial sign-in" token and require the client to complete an MFA challenge (OTP) before issuing the final Access/Refresh tokens.
4.  **Cross-Device QR Login:** A custom session-linking flow where a new device can request a QR challenge, and an already-authenticated mobile device can scan and authorize that session.

### 🛠️ Technical Implementation (Optional)

As a senior architecture plan, here are the exact components and endpoints required for this implementation:

#### 1. Database & Schema Changes (Entity Framework Core)
* **OpenIddict Stores:** Integrate `AddEntityFrameworkCoreStores()` to generate the standard OpenIddict tables (`OpenIddictApplications`, `OpenIddictTokens`, etc.).
* **UserDevices Table:** Create a new entity linked to `.NET Identity User`.
    * Columns: `Id`, `UserId`, `DeviceFingerprintHash` (SHA-256), `DeviceName`, `IsTrusted` (bool), `LastUsed` (DateTime), `CreatedDate`.
* **QRChallenge Cache:** Use Redis (or IDistributedCache) to store short-lived QR challenges. 
    * Key: `ChallengeUUID`, Value: `Status (Pending/Authorized)`, `UserId` (once authorized). TTL: 120 seconds.

#### 2. OpenIddict Server Configuration (`Program.cs`)
* Register OpenIddict Core and Server components.
* Enable **Authorization Code Flow**, **Refresh Token Flow**, and **Client Credentials** (for internal service-to-service if needed).
* Enable Token Rotation: `options.UseRollingRefreshTokens()`.
* Configure exact token lifetimes (e.g., Access Token: 15 mins, Refresh Token: 7 days).
* Register custom event handlers (`HandleTokenRequestContext`) to inject device validation logic into the token pipeline.

#### 3. Core Endpoints
* **`GET /connect/authorize`**: The standard OIDC authorization endpoint. Validates the client and redirects to the login UI.
* **`POST /connect/token`**: Handles the exchange of authorization codes and refresh tokens for access tokens.
* **`GET /connect/logout`**: Handles session termination and token revocation.
* **`GET /connect/userinfo`**: Returns OIDC standard claims (sub, email, name) and custom `amr` (Authentication Methods Reference) claims.

#### 4. Custom Login & Trusted Device Pipeline
* **`POST /api/auth/login`**: 
    * Validates username/password via `.NET 10 SignInManager`.
    * Hashes the incoming `X-Device-Fingerprint` header.
    * Checks the `UserDevices` table. If `IsTrusted == true`, proceed to issue the Authorization Code.
    * If `IsTrusted == false`, generate an MFA OTP, trigger the email/SMS service, and return a `403 Forbidden` with a custom payload: `{ "error": "mfa_required", "partial_token": "<jwt>" }`.
* **`POST /api/auth/verify-mfa`**:
    * Validates the `partial_token` and the user-provided OTP.
    * If `trustDevice == true` in the payload, add the `DeviceFingerprintHash` to the `UserDevices` table.
    * Issue the final Authorization Code to complete the OIDC flow.

#### 5. QR System Implementation
* **`POST /api/qr/generate`**: Generates a UUID and stores it in Redis. Returns the UUID to the unauthenticated client to render as a QR code.
* **`POST /api/qr/authorize`** *(Requires Bearer Token)*: The authenticated mobile app calls this with the scanned UUID. The IdP marks the UUID in Redis as `Authorized` and attaches the mobile user's `UserId`.
* **`GET /api/qr/poll/{uuid}`** (or SignalR Hub): The unauthenticated client listens for status changes. Once the IdP confirms authorization, this endpoint issues the Authorization Code for the client to exchange for tokens.

### ✅ Requirements Checklist

- [x] This is a new feature, not a bug fix.
- [x] I have checked existing issues to ensure this isn't a duplicate.
- [x] This change follows the project's architecture.

### 🧪 Alternatives Considered

1.  **Duende IdentityServer:** Considered, but rejected due to high enterprise licensing costs. OpenIddict provides the same OIDC compliance and greater flexibility completely free and open-source.
2.  **Keycloak/Auth0:** Considered offloading to an external IdP. Rejected because we need fine-grained control over the Trusted Device logic and custom database schemas (specifically the custom QR cross-device authorization logic), which are difficult to implement gracefully in third-party SaaS solutions without messy webhooks.
3.  **Standard .NET Identity Cookies:** Rejected because we need to issue standardized JWTs and Refresh Tokens for multiple clients (Web, Mobile, Desktop) via a centralized OIDC architecture.

### PR DESCRIPTION
**Pull request by TheGresta**
## 📝 Description
This PR implements [brief description].

**Closes #45** 👈 *Mandatory: Replace with the actual issue number*

---

## 🛠️ Type of Change
- [ ] 🚀 New Feature (non-breaking change which adds functionality)
- [ ] 🐛 Bug Fix (non-breaking change which fixes an issue)
- [ ] 🧹 Refactor (code cleanup, no functional changes)
- [ ] 📝 Documentation Update
- [ ] ⚠️ Breaking Change (fix or feature that would cause existing functionality to change)

---

## 🧪 How Has This Been Tested?
- [ ] **Unit Tests:** `npm test` (or equivalent) passed successfully.
- [ ] **Manual Testing:** Describe how you verified this in the browser/app.
- [ ] **Edge Cases:** What happens if the input is empty or the API fails?

---

## 📸 Screenshots / Screen Recordings
| Before | After |
| :--- | :--- |
| [Image/GIF] | [Image/GIF] |

---

## ✅ Checklist
- [ ] My branch follows the naming convention: `type/issue-number-description`.
- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation accordingly.
- [ ] My changes generate no new warnings.

## Summary by CodeRabbit

* **New Features**
  * Added a full Security / Identity Provider (IdP) module with MFA, device trust management, passkey (WebAuthn/FIDO2) support, QR cross-device auth, recovery codes, and session/consent management.
  * Built role-based access control with permissions, role conflicts, and user role assignment.
  * Added audit logging and security event tracking for compliance and forensics.
  * Exposed an initial API surface (includes a sample weather endpoint) and local run profiles for development.

### CODE REVIEW

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Aggregates/AuditAggregate/AuditLog.cs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            ResourceId,
+            IpAddress,
+            CorrelationId,
+            IsSuccess,
+            OccurredAtUtc = OccurredAtUtc.ToString("O"),
+            Payload
+        });
+
+        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(canonical));
+        return Convert.ToHexString(bytes).ToLowerInvariant();
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Aggregates/AuditAggregate/AuditLog.cs]: create me suggestion codes for this

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Common/IRepository.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    void Delete(TEntity entity);
+
+    /// <summary>
+    /// Marks the entity for physical removal from the database.
+    /// </summary>
+    /// <param name="entity">The entity to permanently remove.</param>
+    void HardDelete(TEntity entity);
+
+    #endregion
+}
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Common/IRepository.cs]: create me gold standart codes for base repository so i can refactor it. Also update this base repository for relation between entities and include steps

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Common/IUnitOfWork.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    /// <param name="cancellationToken">
+    /// A <see cref="CancellationToken"/> to observe while waiting for the task to complete. 
+    /// The default is <see langword="default"/>.
+    /// </param>
+    /// <returns>
+    /// A <see cref="Task{TResult}"/> representing the asynchronous operation. The task result contains the 
+    /// number of state entries written to the database.
+    /// </returns>
+    /// <exception cref="DbUpdateException">Thrown if an error occurs while saving to the database.</exception>
+    /// <exception cref="DbUpdateConcurrencyException">Thrown if a concurrency violation is encountered.</exception>
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Common/IUnitOfWork.cs]: create me required updates

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/ExternalLogin.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+        Id = ExternalLoginId.New();
+        UserId = userId;
+        Provider = provider.ToLowerInvariant().Trim();
+        ProviderSubjectId = providerSubjectId.Trim();
+        ProviderEmail = providerEmail?.Trim().ToLowerInvariant();
+        LinkedAt = DateTime.UtcNow;
+    }
+
+    public void RecordUsage() => LastUsedAt = DateTime.UtcNow;
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/ExternalLogin.cs]: create me gold standart codes for this refactoring

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/Permission.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                "Permission name must follow namespace:resource:action format.",
+                "INVALID_PERMISSION_FORMAT");
+
+        Id = id;
+        Name = name.ToLowerInvariant().Trim();
+        Description = description?.Trim() ?? string.Empty;
+        Effect = effect;
+        ResourcePattern = resourcePattern?.Trim();
+        Conditions = conditions;
+    }
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/Permission.cs:62]: give me suggestion updates

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/RecoveryCode.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        IsConsumed = false;
+    }
+
+    internal void Consume()
+    {
+        if (IsConsumed)
+            throw new DomainException("Recovery code has already been used.", "RECOVERY_CODE_CONSUMED");
+
+        IsConsumed = true;
+        ConsumedAt = DateTime.UtcNow;
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/RecoveryCode.cs]: give me suggested fixes for this in gold standart

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/Role.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+        if (!_conflictsWiths.Contains(conflictingRoleId))
+            _conflictsWiths.Add(conflictingRoleId);
+    }
+
+    public void RemoveConflict(RoleId conflictingRoleId) =>
+        _conflictsWiths.Remove(conflictingRoleId);
+
+    public bool ConflictsWith(RoleId roleId) =>
+        _conflictsWiths.Contains(roleId);
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/Role.cs:149]: give me suggested solution in gold standarts

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/User.cs [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+    }
+
+    // ── Domain Behaviours: Phone ──────────────────────────────────────────────
+
+    public void SetPhone(string e164Number)
+    {
+        PhoneNumber = PhoneNumber.Create(e164Number);
+        IsPhoneConfirmed = false;
+        UpdatedAt = DateTime.UtcNow;
+    }
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/User.cs:215]: give me suggested fixes in gold standart

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/UserConsent.cs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        AddDomainEvent(new UserConsentRevokedEvent(UserId, ClientId.Value));
+    }
+
+    public void MarkSuperseded(UserConsentId newConsentId)
+    {
+        Status = ConsentStatus.Superseded;
+        IsActive = false;
+        UpdatedAt = DateTime.UtcNow;
+
+        AddDomainEvent(new UserConsentSupersededEvent(UserId, ClientId.Value, newConsentId));
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/UserConsent.cs]: give me gold standart solution for this fixes

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/UserConsent.cs [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+        foreach (var name in newScopeNames)
+        {
+            var normalized = name.ToLowerInvariant().Trim();
+            if (!_scopes.Any(s => s.Name == normalized))
+                _scopes.Add(new ConsentScope(name));
+        }
+
+        UpdatedAt = DateTime.UtcNow;
+    }
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Entities/UserConsent.cs]: what we need to do here than ? give me gold standart solution

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Repositories/IUserRepository.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+using Playbook.Security.IdP.Domain.Entities;
+using Playbook.Security.IdP.Domain.Entities.Ids;
+
+namespace Playbook.Security.IdP.Domain.Repositories;
+
+public interface IUserRepository : IRepository<User, UserId>
+{
+    Task<User?> GetByEmailAsync(string email, CancellationToken ct = default);
+    Task<User?> GetByUsernameAsync(string username, CancellationToken ct = default);
+    Task<bool> IsEmailUniqueAsync(string email, CancellationToken ct = default);
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Repositories/IUserRepository.cs:11]: suggess me gold standart solution for this

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Services/IPasswordHasher.cs [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+{
+    /// <summary>
+    /// Returns a strongly-typed PasswordHash Value Object.
+    /// </summary>
+    PasswordHash HashPassword(string password);
+
+    /// <summary>
+    /// Accepts the Value Object for verification.
+    /// </summary>
+    bool VerifyPassword(string password, PasswordHash storedHash);
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Services/IPasswordHasher.cs]: give me suggestion gold standart codes for this

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Services/IUniqueIdentityChecker.cs [hunk 13]
```diff
@@ -0,0 +1,20 @@
+﻿using Playbook.Security.IdP.Domain.ValueObjects;
+
+namespace Playbook.Security.IdP.Domain.Services;
+
+/// <summary>
+/// Domain service to ensure global uniqueness of identity identifiers.
+/// This prevents race conditions and duplicate account creation.
+/// </summary>
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Services/IUniqueIdentityChecker.cs]: how can i solve this problem in gold standart

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Specifications/TrustedDeviceSpecification.cs [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+    public Expression<Func<UserDevice, bool>> ToExpression()
+    {
+        var cutoff = DateTime.UtcNow.Subtract(_maxInactivityPeriod);
+
+        return device => device.IsActive &&
+                         device.IsTrusted &&
+                         !device.IsSuspended &&
+                         device.LastUsedAt >= cutoff &&
+                         device.FailureCount < 3;
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Specifications/TrustedDeviceSpecification.cs:34]: give me suggestion solution for this in gold standart

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Application/Behaviors/AuditBehavior.cs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            // We pull data from the IAuditableRequest (The "What") 
+            // and the IRequestContext (The "Who/Where")
+            var auditLog = AuditLog.Create(
+                actorId: requestContext.UserId,
+                action: typeof(TRequest).Name,
+                resourceName: request.ResourceName,
+                resourceId: request.ResourceId,
+                serviceName: string.Empty,
+                environment: string.Empty,
+                hmacKey: [],
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Application/Behaviors/AuditBehavior.cs]: give me gold standart suggestion solution

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Application/Behaviors/IdempotencyBehavior.cs [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                LogIdempotencyHit(logger, requestId);
+                return cachedResponse;
+            }
+
+            // If it's reserved but has no response, it's currently being processed by another thread/node
+            LogIdempotencyConflict(logger, requestId);
+            return (TResponse)(dynamic)Error.Conflict(
+                code: "Idempotency.InProgress",
+                description: "This request is currently being processed. Please wait.");
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Application/Behaviors/IdempotencyBehavior.cs]: give me gold standart solution for this problem

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Application/Behaviors/TransactionBehavior.cs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+            // 2. Dispatch Domain Events BEFORE committing.
+            // If an event handler fails (e.g., trying to assign a role that doesn't exist),
+            // it will throw an exception and trigger the catch block (Rollback).
+            await dispatcher.DispatchEventsAsync(cancellationToken);
+
+            // 3. Final Persistence
+            await unitOfWork.SaveChangesAsync(cancellationToken);
+
+            // 4. Seal the deal
+            await unitOfWork.CommitTransactionAsync(cancellationToken);
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Application/Behaviors/TransactionBehavior.cs:50]: but i also want to spot the transaction if any invalid events happen before the save changes. What is the best practice in gold standarts here. How can i solve this problem. Give me a solution.

File: src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Common/AuditableEntity.cs [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        CreatedAt = createdAt;
+        CreatedBy = createdBy;
+    }
+
+    public void SetUpdateMetadata(DateTime updatedAt, UserId updatedBy)
+    {
+        UpdatedAt = updatedAt;
+        UpdatedBy = updatedBy;
+    }
```
[the remark this answers is not in the document]
Comment (TheGresta) [reply src/Security/Playbook.Security.IdP/Playbook.Security.IdP.Domain/Common/AuditableEntity.cs]: if i update them to private set, sub classes from auditEntity can not access to se this properties. What we need to do here ?
